nixos/modules/services/web-servers/traefik.nix

```diff
@@ -4,56 +4,111 @@ with lib;
 
 let
   cfg = config.services.traefik;
-  configFile =
-    if cfg.configFile == null then
+  jsonValue = with types; let
+    valueType = nullOr (oneOf [
+      bool
+      int
+      float
+      str
+      (lazyAttrsOf valueType)
+      (listOf valueType)
+    ]) // {
+      description = "JSON value";
+      emptyValue.value = {};
+    };
+  in valueType;
+  dynamicConfigFile =
+    if cfg.dynamicConfigFile == null then
       pkgs.runCommand "config.toml" {
         buildInputs = [ pkgs.remarshal ];
         preferLocalBuild = true;
       } ''
         remarshal -if json -of toml \
-          < ${pkgs.writeText "config.json" (builtins.toJSON cfg.configOptions)} \
+          < ${pkgs.writeText "dynamic_config.json" (builtins.toJSON cfg.dynamicConfigOptions)} \
+          > $out
+      ''
+    else cfg.dynamicConfigFile;
+  recursiveMerge = attrList:
+    let f = attrPath:
+      zipAttrsWith (n: values:
+        if tail values == []
+        then head values
+        else
+          if all isList values
+          then unique (concatLists values)
+          else
+            if all isAttrs values
+            then f (attrPath ++ [n]) values
+            else last values
+      );
+    in f [] attrList;
+  staticConfigFile =
+    if cfg.staticConfigFile == null then
+      pkgs.runCommand "config.toml" {
+        buildInputs = [ pkgs.yj ];
+        preferLocalBuild = true;
+      } ''
+        yj -jt -i \
+          < ${pkgs.writeText "static_config.json" (builtins.toJSON (recursiveMerge [cfg.staticConfigOptions { providers.file.filename = "${dynamicConfigFile}";}]))} \
           > $out
       ''
-    else cfg.configFile;
+    else cfg.staticConfigFile;
 
 in {
   options.services.traefik = {
     enable = mkEnableOption "Traefik web server";
 
-    configFile = mkOption {
+    staticConfigFile = mkOption {
       default = null;
-      example = literalExample "/path/to/config.toml";
+      example = literalExample "/path/to/static_config.toml";
       type = types.nullOr types.path;
       description = ''
-        Path to verbatim traefik.toml to use.
-        (Using that option has precedence over <literal>configOptions</literal>)
+        Path to traefik's static configuration to use.
```

Comment:
Is the static/dynamic configuration a concept that traefic has or is it something that NixOS introduces? If it is the latter one, it should be mentioned in the documentation.

Reply:
Static vs Dynamic configuration is a concept Traefik has. Static being that specific config file. And dynamic being for example a label on a container.